File: frontend/src/types/optionTrade.ts
/** Core OptionTrade interface matching backend OptionTradeRead schema. */
export interface OptionTrade {
  id: number
  trade_id: string
  source_trade_id: string | null
  review_id: string | null
  leg: string | null
  portfolio_id: number | null
  counterparty_id: number | null
  portfolio_name: string | null
  counterparty_name: string | null

  // Polymorphic discriminator
  option_category: string

  // Option specs
  option_type: string | null
  trade_type: string | null   // CALL / PUT
  direction: string | null     // 买入 / 卖出
  strike: number | null
  ccy_pair: string | null
  trade_currency: string | null

  // Notional
  notional1: number | null
  notional2: number | null

  // Dates
  trade_date: string | null
  expiry_date: string | null
  delivery_date: string | null
  premium_payment_date: string | null

  // Premium
  premium_type: string | null
  premium_rate: number | null
  premium_amount: number | null
  premium_currency: string | null

  // Market data
  spot_rate: number | null
  volatility: number | null

  // Status
  exercise_status: string | null
  effective_status: string | null
  allocation_status: string | null

  // Venue
  venue: string | null
  clearing_method: string | null

  // Meta
  tenor: string | null
  event_type: string | null
  trade_purpose: string | null
  operator: string | null
  source: string | null
  comments: string | null

  // Timestamps
  created_at: string | null
  updated_at: string | null
}

export interface OptionTradeListResponse {
  data: OptionTrade[]
  total: number
  page: number
  page_size: number
}

export interface OptionTradeFilterParams {
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: string
  portfolio_id?: number | null
  counterparty_id?: number | null
  ccy_pair?: string
  option_type?: string
  trade_type?: string
  direction?: string
  expiry_from?: string
  expiry_to?: string
  trade_date_from?: string
  trade_date_to?: string
  search?: string
  exercise_status?: string
  option_category?: string
}

export interface OptionTradeCreate {
  trade_id: string
  option_category?: string
  source_trade_id?: string | null
  review_id?: string | null
  leg?: string | null
  portfolio_id?: number | null
  portfolio_name?: string | null
  counterparty_name?: string | null
  option_type?: string | null
  trade_type?: string | null
  direction?: string | null
  strike?: number | null
  ccy_pair?: string | null
  trade_currency?: string | null
  notional1?: number | null
  notional2?: number | null
  trade_date?: string | null
  expiry_date?: string | null
  delivery_date?: string | null
  premium_payment_date?: string | null
  premium_type?: string | null
  premium_rate?: number | null
  premium_amount?: number | null
  premium_currency?: string | null
  spot_rate?: number | null
  volatility?: number | null
  exercise_status?: string | null
  effective_status?: string | null
  allocation_status?: string | null
  venue?: string | null
  clearing_method?: string | null
  tenor?: string | null
  event_type?: string | null
  trade_purpose?: string | null
  operator?: string | null
  source?: string | null
  comments?: string | null
}

export interface OptionTradeUpdate {
  trade_id?: string | null
  option_category?: string
  source_trade_id?: string | null
  review_id?: string | null
  leg?: string | null
  portfolio_id?: number | null
  portfolio_name?: string | null
  counterparty_name?: string | null
  option_type?: string | null
  trade_type?: string | null
  direction?: string | null
  strike?: number | null
  ccy_pair?: string | null
  trade_currency?: string | null
  notional1?: number | null
  notional2?: number | null
  trade_date?: string | null
  expiry_date?: string | null
  delivery_date?: string | null
  premium_payment_date?: string | null
  premium_type?: string | null
  premium_rate?: number | null
  premium_amount?: number | null
  premium_currency?: string | null
  spot_rate?: number | null
  volatility?: number | null
  exercise_status?: string | null
  effective_status?: string | null
  allocation_status?: string | null
  venue?: string | null
  clearing_method?: string | null
  tenor?: string | null
  event_type?: string | null
  trade_purpose?: string | null
  operator?: string | null
  source?: string | null
  comments?: string | null
}

File: backend/app/schemas/option_trade.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator


class OptionTradeBase(BaseModel):
    """Common fields shared across option trade schemas."""

    # Polymorphic discriminator
    option_category: str = "fx_vanilla"

    # Core identifiers
    source_trade_id: str | None = None
    review_id: str | None = None
    leg: str | None = None
    portfolio_id: int | None = None
    counterparty_id: int | None = None
    portfolio_name: str | None = None
    counterparty_name: str | None = None

    # Option specs
    option_type: str | None = None
    trade_type: str | None = None
    direction: str | None = None
    strike: float | None = None
    ccy_pair: str | None = None
    trade_currency: str | None = None

    # Notional
    notional1: float | None = None
    notional2: float | None = None

    # Dates
    trade_date: date | None = None
    expiry_date: date | None = None
    delivery_date: date | None = None
    premium_payment_date: date | None = None

    # Premium
    premium_type: Literal["Pips", "%"] | None = None
    premium_rate: float | None = None
    premium_amount: float | None = None
    premium_currency: str | None = None

    # Market data
    spot_rate: float | None = None
    volatility: float | None = None

    # Status
    exercise_status: str | None = None
    delivery_status: str | None = None
    effective_status: str | None = None
    allocation_status: str | None = None

    # Venue
    venue: str | None = None
    clearing_method: str | None = None

    # Meta
    tenor: str | None = None
    event_type: str | None = None
    trade_purpose: str | None = None
    operator: str | None = None
    source: str | None = None
    comments: str | None = None

    @field_validator("premium_currency")
    @classmethod
    def validate_premium_currency(cls, v: str | None, info) -> str | None:
        if v is None:
            return v
        ccy_pair = info.data.get("ccy_pair")
        if ccy_pair:
            parts = ccy_pair.split("/")
            if len(parts) == 2 and v not in parts:
                raise ValueError("premium_currency must be one of the currencies in ccy_pair")
        return v


class OptionTradeRead(OptionTradeBase):
    """Schema for reading an OptionTrade (API response)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    trade_id: str
    created_at: datetime | None = None
    updated_at: datetime | None = None


class OptionTradeCreate(OptionTradeBase):
    """Schema for creating a new option trade manually."""

    trade_id: str


class OptionTradeUpdate(OptionTradeBase):
    """Fields that can be updated on an option trade."""

    trade_id: str | None = None


class OptionTradeListResponse(BaseModel):
    """Paginated list response for option trades."""

    data: list[OptionTradeRead]
    total: int
    page: int
    page_size: int


class OptionTradeFilterParams(BaseModel):
    """Query parameters for filtering option trades list."""

    page: int = 1
    page_size: int = 50
    sort_by: str = "trade_id"
    sort_order: str = "asc"  # asc or desc
    portfolio_id: int | None = None
    counterparty_id: int | None = None
    ccy_pair: str | None = None
    option_type: str | None = None
    trade_type: str | None = None
    direction: str | None = None
    expiry_from: date | None = None
    expiry_to: date | None = None
    trade_date_from: date | None = None
    trade_date_to: date | None = None
    search: str | None = None
    exercise_status: str | None = None
    option_category: str | None = None
